qubesadmin/tools/qvm_template_postprocess.py

```diff
@@ -168,29 +168,36 @@ def post_install(args):
     if not args.skip_start:
         # just created, so no need to save previous value - we know what it was
         vm.netvm = None
-        vm.start()
+        # temporarily enable qrexec feature - so vm.start() will wait for it;
+        # if start fails, rollback it
+        vm.features['qrexec'] = True
         try:
-            vm.run_service_for_stdio('qubes.PostInstall')
-        except qubesadmin.exc.QubesVMError:
-            vm.log.error('qubes.PostInstall service failed')
-        vm.shutdown()
-        if have_events:
-            try:
-                # pylint: disable=no-member
-                qubesadmin.events.utils.wait_for_domain_shutdown(vm,
-                    qubesadmin.config.defaults['shutdown_timeout'])
-            except qubesadmin.exc.QubesVMShutdownTimeout:
-                vm.kill()
-            asyncio.get_event_loop().close()
+            vm.start()
+        except qubesadmin.exc.QubesException:
+            del vm.features['qrexec']
         else:
-            timeout = qubesadmin.config.defaults['shutdown_timeout']
-            while timeout >= 0:
-                if vm.is_halted():
-                    break
-                time.sleep(1)
-                timeout -= 1
-            if not vm.is_halted():
-                vm.kill()
+            try:
+                vm.run_service_for_stdio('qubes.PostInstall')
+            except qubesadmin.exc.QubesVMError:
+                vm.log.error('qubes.PostInstall service failed')
+            vm.shutdown()
+            if have_events:
+                try:
+                    # pylint: disable=no-member
+                    qubesadmin.events.utils.wait_for_domain_shutdown(vm,
+                        qubesadmin.config.defaults['shutdown_timeout'])
+                except qubesadmin.exc.QubesVMShutdownTimeout:
+                    vm.kill()
+                asyncio.get_event_loop().close()
```

Comment:
As to the "standalone tool", I'd imagine someone writing some automatic tool wouldn't want rewrite this function, but just `from qubesadmin.tools.qvm_template_postprocess import post_install`, even if undocumented. This line makes that impossible for no apparent reason.

As to the loop: who started it in the first place? It's used only in `qubesadmin.events.utils.wait_for_domain_shutdown()`, isn't it? If so, I think this is backwards. I'd rewrite `wait_for_domain_shutdown()` *and* `post_install()` as coroutines and `run_` the second one `_until_complete()` from `main()`, then close the loop. IOW, get the loop handling to the outermost context.

I don't think this would cause compat problems, for this package cannot be installed for python<3.4 anyway.